### ISSUE TITLE
Expand lifted custom VJP

### ISSUE THREAD

**Issue by NeilGirdhar** [google/flax#1907]
I'm having two problems with custom VJP:

First, I'm getting "NotImplementedError: This transform does not yet support Modules that include other Modules passed as arguments."

Second, I don't know how (or if it's possible) to return [parameter cotangents](https://github.com/google/flax/discussions/1868).

This is unfortunately a complete blocker for my adoption of Flax.

@jheek Any ideas?

**Comment by jheek** [comment]
You can return parameter cotangents by making the params an explicit return of the forward_fn. e.g.:
```
def forward_fn():

  ...
  return (y, mdl.variables["params"]), res
```

> This transform does not yet support Modules that include other Modules passed as arguments.

This is a bit more tricky. You can try to avoid this by making sure the module that you transform doesn't share Modules passed as an argument to the Module constructor. We currently don't have an easy way to identify which Module is which so we restrict it to just the target Module being transformed.

**Comment by NeilGirdhar** [comment]
Thanks for the quick reply!  Could you elaborate a bit on this? "making sure the module that you transform doesn't share Modules passed as an argument to the Module constructor."   I can't seem to wrap my head around it.

**Comment by NeilGirdhar** [comment]
Also, I'm having trouble connecting what you're suggesting about parameter cotangents to my example.  Let P represent the model variables.
```python
def f(x):
  ...  # This is a function of P

def f_fwd(x):
  return f(x), vjp(...)  # therefore, the VJP returned here depends on P

def f_bwd(v, y_bar):
  x_bar, params_bar = v(y_bar)  # When y_bar is pushed through v, it produces a cotangent of P (params_bar)
  return x_bar  # I need to return params_bar so that the caller of this function receives it as a cotangent.  I don't want to return it as another primal.
```

**Comment by jheek** [comment]
Actually I was confused: Returning params is for defining tangents with respect to params not returning a cotangent.

Let me rephrase to make sure I understand you correctly: When you call custom_vjp you want the return to include the tangents wrt to the variables?

In that case you should just use nn.vjp or nn.jvp in the caller

**Comment by NeilGirdhar** [comment]
> When you call custom_vjp you want the return to include the tangents wrt to the variables?

No.  I tried to explain in my edits.

Essentially, every differentiated Jax function can be interpreted to have a primal output and a cotangent output.   For example,
```python
def f(x):
  ...
  return y

y, x_bar = value_and_grad(f)(y_bar)
```
The primal output is `y`, and the cotangent output is `x_bar`.  I want to write to the parameter cotangent output if possible.  Without using Flax, if `f` accepted weights `w`, I could use a custom VJP to write anything I wanted to w_bar.

**Comment by jheek** [comment]
So if we take the example given in the custom_vjp docstring:
```
class Foo(nn.Module):
      @nn.compact
      def __call__(self, x):
        def f(mdl, x):
          return mdl(x)

        def fwd(mdl, x):
          return nn.vjp(f, mdl, x)

        def bwd(vjp_fn, y_t):
          input_t, params_t = vjp_fn(y_t)
          params_t = jax.tree_map(jnp.sign, params_t)
          return input_t, params_t
```

Here params_t is the parameter cotangent you call params_bar, correct? And in the example it is manipulated such that it is quantized to one bit. Doesn't your problem reduce to replacing the line ` params_t = jax.tree_map(jnp.sign, params_t)` with your own custom logic to overwrite the cotangent?

**Comment by NeilGirdhar** [comment]
@jheek Yes, my apologies!!  I'll give this a shot tomorrow.  It's 5am now and I should probably sleep.

Would you mind illustrating what you mean by "making sure the module that you transform doesn't share Modules passed as an argument to the Module constructor."?  Maybe I'll figure it out tomorrow.

Thanks again!

**Comment by jheek** [comment]
This seems fixed, closing for now
